ChannelRouter and graph re-org awareness

halseth opened the pull request:
This PR makes the channel router re-org aware, by making the router handle stale blocks delivered from the chain view. The channel graph now gets support for "reverse pruning", through the method `DisconnectBlockAtHeight`. This method will remove edges from the channel graph that gets "negative confirmations" because of a reorg, leaving the database in a state consistent with the best height the router is aware of.

### blockEventQueue
This PR also introduces a new queue implementation for ordered block events, modeled after the `htlcswitch`'s `packetQueue`. This is used to ensure the notifications about connected/disconnected blocks from the chain view is handled by the router in the correct order.

### Tests
Several tests are added to ensure the re-org awareness is working as expected in the affected parts of the code:
`router`
`channeldb`
`routing/chainview`
`lnd_test`

~~NB: Needs a fix for https://github.com/lightninglabs/neutrino/pull/18 for all tests to be able to pass.~~ Fixed

Closes #270.

halseth commented on routing/chainview/interface_test.go:
```diff
@@ -451,9 +454,185 @@ func testFilterSingleBlock(node *rpctest.Harness, chainView FilteredChainView,
+		select {
+		case block := <-newBlocks:
+			if i < 5 {
+				t.Fatalf("did not expect to get new block yet %d", i)
```
This was an error in the test, by not consuming the 5 new blocks after mining them. Fixed.

aakselrod replied:
Confirmed that this test is fixed.

aakselrod commented on routing/chainview/interface_test.go:
```diff
@@ -451,9 +454,185 @@ func testFilterSingleBlock(node *rpctest.Harness, chainView FilteredChainView,
+					expectedHeight, block.Height)
+			}
+		case <-time.After(10 * time.Second):
+			t.Fatalf("did not get disconnected block")
```
This part of the test also looks like it's fixed by the change.

halseth commented on routing/chainview/btcd.go:
```diff
@@ -117,6 +115,8 @@ func (b *BtcdFilteredChainView) Start() error {
 
 	b.bestHash, b.bestHeight = *bestHash, bestHeight
 
+	b.blockQueue.Start()
```
Cool, done. I set the filter during an `UpdateFilter` call, I think it is just empty at this point, before calling `NotifyBlocks`.

Roasbeef replied:
Is `NotifyBlocks` in this PR? Working through the commits and haven't seen it _yet_

halseth replied:
It's called from `BtcdFilteredChainView.Start()`

halseth commented:
Addressed comments, should be ready for review again. PTAL.

aakselrod commented on routing/chainview/btcd.go:
```diff
@@ -241,87 +273,115 @@ func (b *BtcdFilteredChainView) chainFilterer() {
 				blockHash, err := b.btcdConn.GetBlockHash(int64(i))
 				if err != nil {
-					log.Errorf("Unable to get block hash: %v", err)
+					log.Errorf("Unable to get block "+
```
It might be nice to provide the user with the height of the block in the error message here. Also, because this is expected in certain reorg conditions, it might be better as a `Warnf` instead of `Errorf` (maybe @Roasbeef can chime in on this).

Roasbeef replied:
Agreed that it would be beneficial to also log the height here!

aakselrod commented on routing/chainview/btcd.go:
```diff
@@ -241,87 +273,115 @@ func (b *BtcdFilteredChainView) chainFilterer() {
+					[]chainhash.Hash{*blockHash})
 				if err != nil {
-					log.Errorf("Unable to get block: %v", err)
+					log.Errorf("Unable to rescan "+
```
The word "blocks" should be singular here, and the hash and/or height of the block might be good to provide to the user in this error message. In addition, since this is expected in certain reorg conditions, it might also be better as a warning.

aakselrod commented on routing/chainview/btcd.go:
```diff
@@ -241,87 +273,115 @@ func (b *BtcdFilteredChainView) chainFilterer() {
+				// If no block was returned from the rescan,
+				// it means no maching transactions were found.
+				if len(rescanned) != 1 {
+					log.Infof("no matching block found for"+
```
This could probably be a `Debugf` instead, since we expect it to happen fairly often.

Roasbeef replied:
^ Ditto

aakselrod commented on routing/chainview/interface_test.go:
```diff
@@ -451,9 +454,211 @@ func testFilterSingleBlock(node *rpctest.Harness, chainView FilteredChainView,
+	}
+
+	// Now we trigger a small reorg, by disconnecting the nodes, mining
+	// a few nodes on each, then connecting them again.
```
Typo: "nodes" should be "blocks."

Roasbeef commented on routing/chainview/queue.go:
```diff
@@ -0,0 +1,148 @@
+// chainView's call to onBlockConnected/onBlockDisconnected to not
+// get blocked, and for the consumer of the block events to always
+// get the events in the correct order.
+type blockEventQueue struct {
```
Excellent addition! We'll likely re-use portions of this abstraction once we also make the `ChainNotifier` properly re-org aware.

Roasbeef commented on routing/chainview/btcd.go:
```diff
@@ -241,87 +249,116 @@ func (b *BtcdFilteredChainView) chainFilterer() {
+				b.filterMtx.Unlock()
+			}
+
+			// Apply the new TX filter to btcd, which will cause
```
Much cleaner now that we can utilize `LoadTxFilter` 👌

When I initially wrote this, the hadn't yet rebased to `btcd` master, so the call wasn't yet available. With the call available now, a more substantial refactoring can take place within the `ChainNotifier` implementations

Roasbeef commented on routing/router.go:
```diff
@@ -345,6 +351,44 @@ func (r *ChannelRouter) syncGraphWithChain() error {
+			return err
+		}
+	}
+
```
The logging statement below should be modified to use `mainBlockHash` and height instead, as in the case we re-org'd back, the statement wouldn't be accurate.

halseth replied:
I don't think it's necessary, since if we reorg'd back, then `pruneHash == mainBlockHash` and same for height, which should make it accurate.

If I'm wrong, the probably the following loop also would need to change, as it loops from `pruneHeight` to `bestHeight`.

Roasbeef replied:
Nope you're correct, I missed that both `pruneHash` and `pruneHeight` were re-assigned within the loop above.

Roasbeef commented on routing/router.go:
```diff
@@ -442,6 +486,35 @@ func (r *ChannelRouter) networkHandler() {
+			r.routeCache = make(map[routeTuple][]*Route)
+			r.routeCacheMtx.Unlock()
+
+			// TODO(halseth): notify client about the reorg?
```
Which client? Notification clients?

halseth replied:
Yeah, as for new blocks:
```go
// Notify all currently registered clients of the newly
// closed channels.
closeSummaries := createCloseSummaries(blockHeight, chansClosed...)
r.notifyTopologyChange(&TopologyChange{
        ClosedChannels: closeSummaries
})
```

Roasbeef replied:
Ah yes, that should be done. This can be done isn a follow up PR/commit as this one is nearly ready to be merged.

Roasbeef commented on channeldb/graph.go:
```diff
@@ -657,40 +672,137 @@ func (c *ChannelGraph) PruneGraph(spentOutputs []*wire.OutPoint,
+// will "rewind" the graph back to the height below, deleting channels
+// that are no longer confirmed from the graph. The prune log will be
+// set to the last prune height valid for the remaining chain.
+func (c *ChannelGraph) DisconnectBlockAtHeight(height uint32) error {
```
It'd be nice if this function returned all the channels that were _pruned_ just as the old `PruneGraph` does. This way we can notify to any of the the notification clients in the `ChannelRouter` that select channels were re-org'd out of the chain.

halseth commented on routing/router.go:
```diff
@@ -345,6 +351,44 @@ func (r *ChannelRouter) syncGraphWithChain() error {
+			case err == channeldb.ErrGraphNeverPruned:
+				break
+			case err == channeldb.ErrGraphNotFound:
+				break
```
Note to self: I suspect this does not break out of the loop 😨

halseth commented on channeldb/graph_test.go:
```diff
@@ -354,6 +355,193 @@ func TestEdgeInsertionDeletion(t *testing.T) {
+		ChainHash:   key,
+		NodeKey1:    node1.PubKey,
+		NodeKey2:    node2.PubKey,
+		BitcoinKey1: node1.PubKey,
```
Created an internal helper method for this testcase, as for other testcases too many parameters are different for it to make sense.

halseth commented:
Addressed comments + fixed a bug where I would break out of the `switch` instead of the `loop` in `syncChannelWithGraph`. I changed it so that we just `return nil` at this point, since that's what is done if you the graph is in the "not-pruned" state at the start of the method. 

PTAL

Roasbeef commented on lnd_test.go:
```diff
@@ -387,6 +387,184 @@ func testBasicChannelFunding(net *networkHarness, t *harnessTest) {
+	}
+
+	ctxt, _ = context.WithTimeout(ctxb, timeout)
+	closeChannelAndAssert(ctxt, t, net, net.Alice, chanPoint, false)
```
The `miner` instance created here should be shutdown as it's no longer used in any tests after this one (needed since there's also no code to remove _all_ the active btcd harnesses at the end of the integration test run as we do with all active lnd nodes).

halseth replied:
added a `defer miner.TearDown()` right after `SetUp`.

Roasbeef commented on routing/router.go:
```diff
@@ -374,12 +374,13 @@ func (r *ChannelRouter) syncGraphWithChain() error {
+				return nil
 			case err == channeldb.ErrGraphNotFound:
-				break
+				return nil
```
Actually, it's fine as it is since this replicates the same behavior as if the graph was never pruned.
